src/mpi/init/init_async.c

```diff
@@ -59,22 +74,163 @@ static void progress_fn(void *data)
 
+#if defined(MPL_HAVE_MISC_GETNPROCS)
+
+    have_cliques = MPIR_pmi_has_local_cliques();
+    apply_affinity = MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY &&
+        strlen(MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY) > 0;
+
+    if (apply_affinity) {
+        /* Consider nodemap cliques when using debugging CVARs */
+        if (have_cliques) {
+            fprintf(stderr,
+                    "Setting affinity for progress threads cannot work correctly with MPIR_CVAR_NUM_CLIQUES or MPIR_CVAR_ODD_EVEN_CLIQUES.\n");
+        }
+
+        global_rank = MPIR_Process.rank;
+        local_rank =
+            (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->node_comm->rank : 0;
+        if (have_cliques) {
+            /* If local cliques > 1, using local_size from node_comm will have conflict on thread idx.
+             * In multiple nodes case, this would cost extra memory for allocating thread affinity on every
+             * node, but it is okay to solve progress thread oversubscription. */
+            local_size = MPIR_Process.comm_world->local_size;
+        } else {
+            local_size =
+                (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->
+                node_comm->local_size : 1;
+        }
+
+        async_threads_per_node = local_size;
+        thread_affinity = (int *) MPL_malloc(async_threads_per_node * sizeof(int), MPL_MEM_OTHER);
+
+        MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                        (MPL_DBG_FDEST,
+                         " global_rank %d, local_rank %d, local_size %d, async_threads_per_node %d",
+                         global_rank, local_rank, local_size, async_threads_per_node));
+
+        mpi_errno = MPIDI_parse_progress_thread_affinity(thread_affinity, async_threads_per_node);
+        MPIR_ERR_CHKANDJUMP(mpi_errno, mpi_errno, MPI_ERR_OTHER, "**parse_thread_affinity");
+        if (MPIR_Process.rank == 0) {
+            int th_idx;
+            for (th_idx = 0; th_idx < async_threads_per_node; th_idx++) {
+                MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                                (MPL_DBG_FDEST, "affinity: thread %d, processor %d",
+                                 th_idx, thread_affinity[th_idx]));
+            }
+        }
+    }
+#endif
+
     int err = 0;
     MPID_Thread_create((MPID_Thread_func_t) progress_fn, NULL, &progress_thread_id, &err);
     MPIR_ERR_CHKANDJUMP1(err, mpi_errno, MPI_ERR_OTHER, "**mutex_create", "**mutex_create %s",
                          strerror(err));
 
+#if defined(MPL_HAVE_MISC_GETNPROCS)
+    if (apply_affinity) {
+        if (have_cliques) {
+            /* In this case, procs on one physical node are partitioned into different virtual nodes,
+             * global_rank should be used to avoid binding progress threads from different ranks to the same core. */
+            affinity_idx = global_rank;
+        } else {
+            affinity_idx = local_rank;
+        }
```

Comment:
Wrap this logic into the same static function I mentioned above. That function can return the affinity that can be directly passed to `MPL_thread_set_affinity`.

Reply:
Done